include/aws/http/private/h2_connection.h

```diff
@@ -62,26 +63,41 @@ struct aws_h2_connection {
          * When queue is empty, then we send DATA frames from the outgoing_streams_list */
         struct aws_linked_list outgoing_frames_queue;
 
+        /* Maps stream-id to aws_h2_stream_closed_when.
+         * Contains data about streams that were recently closed by this end (sent RST_STREAM frame or END_STREAM flag),
+         * but might still receive frames that remote peer sent before learning that the stream was closed.
+         * Entries are removed after a period of time. */
+        struct aws_hash_table closed_streams_where_frames_might_trickle_in;
+
     } thread_data;
 
-    /* Any thread may touch this data, but the lock must be held */
+    /* Any thread may touch this data, but the lock must be held (unless it's an atomic) */
     struct {
+        /* For checking status from outside the event-loop thread. */
+        struct aws_atomic_var is_open;
+
+        /* If non-zero, reason to immediately reject new streams. (ex: closing) */
+        struct aws_atomic_var new_stream_error_code;
+
         struct aws_mutex lock;
 
         /* New `aws_h2_stream *` that haven't moved to `thread_data` yet */
         struct aws_linked_list pending_stream_list;
 
-        /* If non-zero, reason to immediately reject new streams. (ex: closing) */
-        int new_stream_error_code;
-
         bool is_cross_thread_work_task_scheduled;
 
-        /* For checking status from outside the event-loop thread. */
-        bool is_open;
-
     } synced_data;
 };
 
+/**
+ * The action which caused the stream to close.
+ */
+enum aws_h2_stream_closed_when {
```

Comment:
🤷‍♂️
fun fact: You know I've spent at least 5 minutes trying out different words for this Since the values all sounded good with "when" I stuck with it for the type name.
It's not a public enum, I think it's fine